### PR TITLE
New sources model for carp

### LINKED ISSUE #34: Easier switch for BearerKind
Currently the codebase assumes you're connecting to a local node. If you're connecting to a remote node (TCP connection,) you can change the `BearerKind` to TCP in the code.

It's not great to force people to change code for what should be a config change.

### LINKED ISSUE #35: Improve .env setup
Currently, the `.env` file contains two lines that aren't proper .env syntax:

```bash
SOCKET="$(realpath ../cardano-node/bin/\\.pipe\\cardano-node)"
PGPASSFILE="$(realpath secrets/.pgpass)"
```

The fact we have these is unfortunate because it means you can't just load the dotenv from inside Rust (or Typescript) and instead have to set the env variables by calling the .env through bash.

### LINKED ISSUE #45: Allow block sources other than Oura
Right now Carp uses Oura to fetch the block cbor. However, there is no reason that things should be hard-coded to Oura. You could use any tool that gets you the block cbor such as maybe scrolls or something else that may fetch blocks faster.

One way we could do this is by adding a [tower](https://docs.rs/tower/latest/tower/) implementation for Oura so that Carp can use any task that implements this trait. Doing this, however, probably depends on solving https://github.com/dcSpark/carp/issues/34#issuecomment-1140540177 first

### LINKED ISSUE #47: Improve error messages on missing files
If you have file missings, the errors are really generic and don't help much figuring out which files was missing. This seems like an open issue with Rust itself. Until they improve the missing file error message, we can add our own error message.

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Os { code: 2, kind: NotFound, message: "No such file or directory" })', tasks/src/execution_plan.rs:7:49
```

### PR DESCRIPTION
**Pull request by gostkin**
Closes #34 
Closes #35 
Closes #45 
Closes #47 
Tasks left:
- [x] Crate source / sink structures
- [x] Create engine
- [x] Adapt tasks to work with generalized format
- [x] Make oura source  
- [x] Make cardano sink 
- [x] Create new config format for the indexer 
- [x] Test that sync works correctly (wip)
- [x] Add new configs

### CODE REVIEW

File: indexer/src/perf_aggregator.rs [hunk 1]
```diff
@@ -16,6 +17,14 @@ impl PerfAggregator {
             overhead: Duration::new(0, 0),
         }
     }
+
+    pub fn reset(&mut self) {
+        self.block_parse = Duration::new(0, 0);
+        self.block_fetch = Duration::new(0, 0);
+        self.rollback = Duration::new(0, 0);
+        self.overhead = Duration::new(0, 0);
```
Comment (ecioppettini) [inline indexer/src/perf_aggregator.rs:25]:
```suggestion
        *self = Self::new();
```

File: indexer/tasks/src/dsl/database_task.rs [hunk 2]
```diff
@@ -12,38 +12,41 @@ pub struct BlockGlobalInfo {
     pub epoch_slot: Option<u64>,
 }
 
-pub type BlockInfo<'a, BlockType> = (
+pub type BlockInfo<'a, BlockType, BlockExtraType> = (
```
Comment (ecioppettini) [inline indexer/tasks/src/dsl/database_task.rs:15]: what's this for? is it for supporting different blockchains?

is there a case where we would have the same `BlockType` but a different `BlockExtraType`?

File: indexer/src/main.rs [hunk 3]
```diff
@@ -48,60 +102,43 @@ async fn main() -> anyhow::Result<()> {
...
+    let config: Config = serde_yaml::from_reader(file).with_context(|| {
+        format!(
+            "Cannot read config file {path}",
+            path = config_path.display()
+        )
+    })?;
+
+    let mut sink: CardanoSink = match config.sink {
+        SinkConfig::Cardano { .. } => CardanoSink::new(config.sink, exec_plan).await?,
+        _ => todo!("not supported yet"),
```
[the remark this answers is not in the document]
Comment (gostkin) [reply indexer/src/main.rs:126]: it's single, can be extended in the future

File: indexer/tasks/src/dsl/database_task.rs (hunk 2 above)
Comment (gostkin) [reply indexer/tasks/src/dsl/database_task.rs:15]: for every block type there can be just one block extra type

File: indexer/src/main.rs [hunk 4]
```diff
@@ -1,33 +1,87 @@
...
+#[derive(Debug, Clone, Deserialize)]
+#[serde(tag = "type", rename_all = "snake_case")]
+#[serde(deny_unknown_fields)]
+pub enum SourceConfig {
+    Oura {
+        network: String,
+        socket: String,
+        bearer: BearerKind,
+    },
+    DirectSource {},
```
[the remark this answers is not in the document]
Comment (gostkin) [reply indexer/src/main.rs:70]: rename it to N2N or CardanoSource when you have it

File: indexer/src/common.rs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+                ..
+            } => Some(Point::BlockHeader {
+                slot_nb: *block_slot,
+                hash: block_hash.clone(),
+            }),
+        }
+    }
+}
+
+impl EventObject for CardanoEventType {
```
[the remark this answers is not in the document]
Comment (gostkin) [reply indexer/src/common.rs:43]: you have any suggestions where to put it? we usually can get point to sync from an event, if we're making generic services you need some abstraction to allow that

File: indexer/src/main.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ecioppettini) [reply indexer/src/main.rs:126]: But I mean, rust checks that already, and if you add another one it won't compile.

I think this just makes it harder to figure out what to do if you want to add a new sink. It's not a serious issue, but I also don't see what's the benefit of adding that. I would personally avoid adding wildcard patterns wherever is possible.

File: indexer/src/sources/oura_source.rs [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+#[async_trait::async_trait]
+impl dcspark_blockchain_source::Source for OuraSource {
+    type Event = CardanoEventType;
+    type From = Point;
+
+    /// note: from is ignored here since oura is set up just once
+    async fn pull(&mut self, _from: &Self::From) -> anyhow::Result<Option<Self::Event>> {
+        let input = self
+            .input
+            .recv()
```
Comment (ecioppettini) [inline indexer/src/sources/oura_source.rs:92]: AFAIK this is a blocking function, which doesn't play nice with the async interface. Maybe we should consider using `tokio::task::spawn_blocking` for this? Or a helper thread with an async channel...

File: indexer/src/main.rs [hunk 7]
```diff
@@ -48,60 +102,43 @@ async fn main() -> anyhow::Result<()> {
...
-
-    for handle in handles {
-        handle.join().map_err(|_| anyhow!(""))?;
-    }
+    engine
+        .fetch_and_process(start_from.first().unwrap().clone())
+        .await?;
 
-    Ok(())
+    engine.stop().await
```
Comment (ecioppettini) [inline indexer/src/main.rs:143]: I think this doesn't make a lot of sense. The loop in `fetch_and_process` doesn't exit unless there is an error (would be nice to have the `!` type for that), and if there is an error then the `stop` function will not run.

**Pull request merged**